SI-6642 Adds iteratorFrom, keysIteratorFrom, and valuesIteratorFrom

JamesIry opened the pull request:
SI-6642 Adds iteratorFrom, keysIteratorFrom, and valuesIteratorFrom

Adds the ability to efficiently create an iterator that starts at a
given key or element of a sorted set or map. Similar work is done for
key and value only iterators on maps.

The heart of the work is in RedBlackTree. Most of the rest is pushing
the new api methods throughout the appropriate spots in the collection
API.

Along the way mutable.TreeSet is reimplemented in terms of
RedBlackTree to eliminate code duplication and improve efficiency.

review @adriaanm

JamesIry committed:
SI-6642 Code cleanup on RedBlackTree#TreeIterator

In anticipation of some work needed to implement iteratorFrom, this
commit does some variable renaming and general code clean up on
RedBlackTree's TreeIterator.

JamesIry committed:
SI-6642 Adds iteratorFrom, keysIteratorFrom, and valuesIteratorFrom

Adds the ability to efficiently create an iterator that starts at a
given key or element of a sorted set or map. Similar work is done for
key and value only iterators on maps.

The bulk of the work is in RedBlackTree. Most of the rest is pushing
the new api methods throughout the appropriate spots in the collection
API.

This commit leaves undone some similar work possible on mutable TreeSets

JamesIry committed:
SI-6642 Refactor mutable.TreeSet to use RedBlackTree instead of AVL

There was no reason to have mutable.TreeSet use AVLTree while
immutable.TreeSet and immutable.HashSet used RedBlackTree. In
particular that would have meant duplicating the iteratorFrom logic
unnecessarily. So this commit refactors mutable.TreeSet to use
RedBlackTree for everything, including iteratorFrom. It also adds
a test to make sure TreeSet works as expected.

AVLTree should be dead code since it's private[scala.collection.mutable]
and only used by mutable.TreeSet, but to be safe it's only deprecated
in this commit.

soc commented:
Using plural for the methods `keysIteratorFrom` and `valuesIteratorFrom` sounds wrong because `Iterator` already conveys that potentially multiple elements are involved. What's the reasoning behind it?

JamesIry commented:
Because we already have "keysIterator" and "valuesIterator". I'm just adding "from" variants in parallel with "iterator" => "iteratorFrom"

soc commented:
Ugh ... who did that naming? :-/

JamesIry commented:
Open a ticket. We can debate the merits of a rename (with deprecated forwarders) there.

JamesIry committed:
SI-6642 Code cleanup from review of iteratorFrom

This commit is code cleanup from the review on
https://github.com/scala/scala/pull/2119

* Changes several instances of
    def f[A, ]… to
    def f[A]…
* Changes several instances of
    def f[A](…)(implicit ordering : Ordering[A])… to
    def f[A: Ordering](…)…
* Changes one instance of
    x == null to
    x eq null
* Changes two instances of
    id + bottomid to
    bottomid + id

adriaanm commented on src/library/scala/Enumeration.scala:
```diff
@@ -254,7 +254,8 @@ abstract class Enumeration (initial: Int) extends Serializable {
     def + (value: Value) = new ValueSet(nnIds + (value.id - bottomId))
     def - (value: Value) = new ValueSet(nnIds - (value.id - bottomId))
-    def iterator = nnIds.iterator map (id => thisenum.apply(id + bottomId))
+    def iterator = nnIds.iterator map (id => thisenum.apply(bottomId + id))
```
I guess my original comment was mainly about removing the unnecessary `.apply`, but I'll take it as-is.

adriaanm commented:
LGTM

future work:
- do we have a ticket that schedules the removal of AVLTree?
- the tests are nice and mostly self-explanatory, but some comments there would still be good

JamesIry commented:
https://issues.scala-lang.org/browse/SI-7155
